code/controllers/subsystem/map_gen.dm

```diff
@@ -0,0 +1,168 @@
+	/// The subsystem will, upon firing, run the first subsystem in this list until the list is empty or
+	/// it runs out of time, at which point it will stop. The datum carries data with it that tells the subsystem
+	/// what to do to its turfs next. Map generators are inserted with a "priority", and
+	/// identical priority effectively has its own queue that is handled after all foregoing queues have been emptied.
+	var/list/jobs = list()
+
+/datum/controller/subsystem/map_gen/stat_entry(msg)
+	msg = "Q:[length(jobs)]"
+	return ..()
+
+/datum/controller/subsystem/map_gen/Initialize(timeofday)
+	if(initialized)
+		return
+	if(length(jobs))
+		// This should be set by default anyway. Just in case.
+		can_fire = TRUE
+	initialize_biomes()
+	return ..()
+
+///Initialize all biomes, assoc as type || instance
+/datum/controller/subsystem/map_gen/proc/initialize_biomes()
+	PRIVATE_PROC(TRUE)
+	for(var/biome_path in subtypesof(/datum/biome))
+		var/datum/biome/biome_instance = new biome_path()
+		biomes[biome_path] = biome_instance
+
+/datum/controller/subsystem/map_gen/proc/queue_generation(priority, datum/map_generator/gen_to_queue)
+	if(gen_to_queue.priority)
+		CRASH("SSmap_gen told to queue map_generator ([REF(gen_to_queue)], [gen_to_queue.type]) despite it already being queued at priority [gen_to_queue.priority]!")
+	gen_to_queue.priority = priority
+	// Inserts the map generator into our queue at the correct position for its priority.
+	// BINARY_INSERT inserts after all elements of the same priority in the list, so the list functions as an ordered queue.
+	BINARY_INSERT(gen_to_queue, jobs, /datum/map_generator, gen_to_queue, priority, COMPARE_KEY)
+
+	var/message = "MAPGEN: map_generator datum [get_log_string(gen_to_queue)] added to map generator queue at position [jobs.Find(gen_to_queue)]"
+	log_shuttle(message)
+	log_world(message)
+
+	// if the subsystem is disabled due to running through its previous queue, reenable it
+	can_fire = TRUE
+
+/datum/controller/subsystem/map_gen/fire(resumed = 0)
+	if(!jobs)
+		// in case our only job was removed early due to qdeletion
+		can_fire = FALSE
+		return
+
+	// jobs are sorted by priority, and handled in order
+	for(var/datum/map_generator/cur_map_gen in jobs) // serves as a queue
+		while(cur_map_gen.phase != MAPGEN_PHASE_FINISHED)
+			switch(cur_map_gen.phase)
+				if(MAPGEN_PHASE_GENERATE)
+					handle_gen_phase(cur_map_gen)
+					if(state != SS_RUNNING)
+						return
+				if(MAPGEN_PHASE_AFTERCHANGE)
+					handle_afterchange_phase(cur_map_gen)
+					if(state != SS_RUNNING)
+						return
+				if(MAPGEN_PHASE_RUIN_PLACE)
+					// this phase is a little weird, because placing a map template calls CHECK_TICK.
+					// this is undesirable. although sleeping subsystem fires are technically allowed,
+					// slept procs may wake before the MC, causing them to eat entire ticks for themselves.
+					// thus, ruin placement is handled by a waitfor = FALSE proc that sets the mapgen phase twice
```

Comment:
This... doesnt change anything. Waitfor just means it wont sleep the whole subsystem, it will still wake up and do stuff before the MC can do anything. It should be using MC_TICK_CHECK.